inform7/Internal/Inter/Architecture32Kit/Sections/Glk.i6t

```diff
@@ -284,120 +221,378 @@ Functions for minimal Glk windows support.
-	}
+=
+[ GLK_EVENT_TY_From_Struct ev event_struct
+		evtype win_obj;
+	! Erase this event
+	GLK_EVENT_TY_Destroy(ev);
+
+	! Check we have a known window
+	evtype = event_struct-->0;
+	win_obj = FindGlkWindowFromRefNum(event_struct-->1);
+	if (win_obj == nothing && evtype == evtype_CharInput or evtype_LineInput or evtype_MouseInput or evtype_Hyperlink) {
+		IssueRTP("EventUnknownWindow", "Event is on unknown Glk window.", Architecture32KitRTPs);
+		return ev;
+	}
+
+	! Fill in the event
+	ev-->GLK_EVENT_TYPE_SF = evtype;
+	ev-->GLK_EVENT_WINDOW_SF = win_obj;
+	ev-->GLK_EVENT_VALUE1_SF = event_struct-->2;
+	ev-->GLK_EVENT_VALUE2_SF = event_struct-->3;
+
+	! Event maintenance
+	switch (evtype) {
+		evtype_CharInput:
+			win_obj.text_input_status = INPUT_STATUS_NONE;
+		evtype_LineInput:
+			win_obj.line_input_buffer_curlen = ev-->GLK_EVENT_VALUE1_SF;
+			win_obj.text_input_status = INPUT_STATUS_NONE;
+		evtype_MouseInput:
+			win_obj.requesting_mouse = false;
+		evtype_Hyperlink:
+			win_obj.requesting_hyperlink = false;
+	}
+	return ev;
+];
+
+@ |GLK_EVENT_TY_To_Struct| then does the reverse, applying a Glk event BV back onto
+the event struct.
 
-	! Copy back to the original event structure
-	event_struct-->0 = Glk_Event_Struct_type;
-	if (Glk_Event_Struct_win) {
-		event_struct-->1 = GProperty(OBJECT_TY, Glk_Event_Struct_win, glk_ref);
+=
+[ GLK_EVENT_TY_To_Struct ev event_struct;
+	event_struct-->0 = ev-->GLK_EVENT_TYPE_SF;
+	if (ev-->GLK_EVENT_WINDOW_SF) {
+		event_struct-->1 = (ev-->GLK_EVENT_WINDOW_SF).glk_ref;
 	}
 	else {
 		event_struct-->1 = 0;
 	}
-	event_struct-->2 = Glk_Event_Struct_val1;
-	event_struct-->3 = Glk_Event_Struct_val2;
-	rfalse;
+	event_struct-->2 = ev-->GLK_EVENT_VALUE1_SF;
+	event_struct-->3 = ev-->GLK_EVENT_VALUE2_SF;
 ];
 
-@ |SetGlkEventType| helps us change the current event type, by ensuring that
+@ |GLK_EVENT_TY_Process| applies a glk event onto the current glk event, and
+tells glk_select to re-run the glk event handling rules. It also ensures that
 if there are any pending keyboard input requests they will be cancelled if the
-new event is a keyboard event. It does not however do anything if no keyboard
-input requests are pending, which means the new event could be ignored.
+new event is a keyboard event.
 
 =
-[ SetGlkEventType type;
-	if (type == Glk_Event_Struct_type) {
-		return;
-	}
-	if (type == evtype_CharInput or evtype_LineInput) {
-		if (Glk_Event_Struct_win) {
-			if (Glk_Event_Struct_win.text_input_status == INPUT_STATUS_ACTIVE_CHAR or INPUT_STATUS_ACTIVE_CHAR_UNI) {
-				glk_cancel_char_event(Glk_Event_Struct_win.glk_ref);
-			}
-			else if (Glk_Event_Struct_win.text_input_status == INPUT_STATUS_ACTIVE_LINE or INPUT_STATUS_ACTIVE_LINE_UNI) {
-				glk_cancel_line_event(Glk_Event_Struct_win.glk_ref);
-			}
-			Glk_Event_Struct_win.text_input_status = INPUT_STATUS_NONE;
+Array current_glk_event --> [ BLK_BVBITMAP_SBONLY; GLK_EVENT_TY; 0; 0; 0; 0; 0; 0; 0; ];
```

Comment:
I was asking about globals from I6 because I need to access one like this. But it was easy enough to just manually create like this.